### PR TITLE
support etcd credentials in etcd plugin

### LINKED ISSUE #2441: Support etcd username and password for etcd plugin
Following the discussion from https://github.com/coredns/coredns/pull/2163
I would like to had the support for username and password in the etcd configuration.

In the PR discussion @miekg said he would prefer to support this from the config file instead of some environment variable.

I can work on that and provide a PR that implement that feature.
Is there any other recommendation regarding this feature ?

About the format of the config file, I was thinking implementing something like
```
etcd [zone]{
   endpoint http://localhost:2379
   credentials username password
}
```
I can also be more explicity and do 
```
etcd [zone]{
   endpoint http://localhost:2379
   username username
   password password
}
```
What do you think ?
```
```

### PR DESCRIPTION
**Pull request by zaibon**
fixes #2441

### CODE REVIEW

File: plugin/etcd/README.md [hunk 1]
```diff
@@ -113,6 +113,15 @@ etcd skydns.local {
 ...
 ~~~
 
+You can connect to an etcd cluster protected by a password by specifiying the credential in the Corefile.
+
+~~~
+etcd skydns.local {
+    endpoint http://localhost:2379
+    credentials username password
```
Comment (miekg) [inline plugin/etcd/README.md]: can you add this to the initial how-to-use the etcd plugin, with:

~~~
credentials USERNAME PASSWORD
~~~~
and explain it's use in the bullets below?

File: plugin/etcd/README.md [hunk 2]
```diff
@@ -213,4 +222,4 @@ If you query the zone name for `TXT` now, you will get the following response:
 ~~~ sh
 % dig +short skydns.local TXT @localhost
 "this is a random text message."
-~~~
+~~~
```
Comment (miekg) [inline plugin/etcd/README.md:213]: what's this change? Fixing trailing whitespace?

**Review by miekg — COMMENTED** [review]
is there a test that's easy to add?

File: plugin/etcd/README.md (hunk 2 above)
Comment (zaibon) [reply plugin/etcd/README.md:213]: Yep looks like it

File: plugin/etcd/README.md (hunk 1 above)
Comment (zaibon) [reply plugin/etcd/README.md]: Which "how-to-use" section are you talking about?

**Comment by zaibon** [comment]
> is there a test that's easy to add?
Do you have some test about the parsing of the config file already somewhere? I didn't really find one.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] support etcd ..." ]
>> is there a test that's easy to add?
>Do you have some test about the parsing of the config file already somewhere? I didn't really find one.

test/etcd_test.go has tests.

If you're adding some please *don't* use proxy.Lookup as I'm removing all the 
code: new way of doing a lookup: https://github.com/coredns/coredns/pull/2436

tx

File: plugin/etcd/README.md (hunk 1 above)
Comment (miekg) [reply plugin/etcd/README.md]: starting on line 30

Comment (zaibon) [reply plugin/etcd/README.md]: Done

**Comment by zaibon** [comment]
I think the PR is ready now

File: plugin/etcd/README.md [hunk 3]
```diff
@@ -45,6 +46,7 @@ etcd [ZONES...] {
   queries for those zones will be subject to fallthrough.
 * **PATH** the path inside etcd. Defaults to "/skydns".
 * **ENDPOINT** the etcd endpoints. Defaults to "http://localhost:2379".
+* `credentials` If the etcd cluster pointed by `endpoint` is password protectd, use this field to specify the username and password.
```
Comment (miekg) [inline plugin/etcd/README.md]: protectd -> protected

better will be:
~~~
`credentials` is used to set the **USERNAME** and **PASSWORD** for accessing the etcd cluster.
~~~

File: plugin/etcd/README.md [hunk 4]
```diff
@@ -113,6 +115,7 @@ etcd skydns.local {
 ...
 ~~~
 
+
```
Comment (miekg) [inline plugin/etcd/README.md]: unintended newline?

File: plugin/etcd/setup.go [hunk 5]
```diff
@@ -103,6 +105,15 @@ func etcdParse(c *caddy.Controller) (*Etcd, bool, error) {
 					if err != nil {
 						return &Etcd{}, false, err
 					}
+				case "credentials":
+					args := c.RemainingArgs()
+					if len(args) == 0 {
+						return &Etcd{}, false, c.ArgErr()
+					}
+					if len(args) != 2 {
+						return &Etcd{}, false, c.Errf("credentials required 2 arguments, username and password")
```
Comment (miekg) [inline plugin/etcd/setup.go]: required -> requires

File: plugin/etcd/setup_test.go [hunk 6]
```diff
@@ -14,43 +14,69 @@ func TestSetupEtcd(t *testing.T) {
...
 		},
 		// negative
 		{
 			`etcd {
 	endpoints localhost:300
 }
-`, true, "", []string{""}, "unknown property 'endpoints'",
+`, true, "", []string{""}, "unknown property 'endpoints'", "", "",
+		},
+		// with valid crendentials
```
Comment (miekg) [inline plugin/etcd/setup_test.go]: credentials

**Comment by miekg** [comment]
yes, starting to look good, but def. needs a test; I can be sure if this works or not.

In the top-level test/ dir you can copy etcd_cache_test.go and make a minimal etcd test that sets a username and password on the etcd backend (and deletes that again - because the testing etcd is used for more tests) and queries a single record from etcd and sees if it correct.

**Comment by zaibon** [comment]
@miekg can you have another and see if https://github.com/coredns/coredns/pull/2442/files#diff-de7cfcdf070628c0ec52593ecfda6801R12 is what you are expecting for the tests.
I still need to figure out how to disable authentication to clean up the test. Everything Ive tried so far returns an error. If you have some input about that.

**Comment by miekg** [comment]
you mean this code:
~~~// defer func() {
	// 	if _, err := etc2.Client.AuthDisable(ctx); err != nil {
	// 		t.Errorf("Fail to disable authentication: %v", err)
	// 	}
	// 	if _, err := etc.Client.RoleDelete(ctx, "root"); err != nil {
	// 		t.Errorf("Failed to delete root role: %v", err)
	// 	}
	// 	_, err := etc.Client.UserDelete(ctx, username)
	// 	if err != nil {
	// 		t.Errorf("Failed to delete user: %v", err)
	// 	}
	// }()
~~~
What happens?

Otherwise that test looks like the thing I want there

**Comment by zaibon** [comment]
@miekg ok finally all tests are green.
Problem was I tried to remove root user and role, while etcd doesn't allow it.

**Review by miekg — APPROVED** [review]

**Pull request merged**